### ISSUE TITLE
Separate Dependencies for Development and Production

### ISSUE THREAD

**Issue by crantila** [Abjad/abjad#468]
When someone installs Abjad with ``pip``, they install all of the dependencies listed in the ``install_requires`` variable in ``setup.py``. This includes ``Sphinx``, which is used to build documentation from program code, and ``pytest``, which is used to run the automated test suite. While ``pytest`` is quite small, ``Sphinx`` is a large package and it requires several large-ish dependencies.

Since these packages are (or should be) needed only for Abjad development, I suggest removing them from the ``install_requires`` argument, relieving others from having to install this software that will go unused.

Of course, I may be missing something, in which case I'd be happy to learn. Thanks!

**Comment by josiah-wolf-oberholtzer** [comment]
Somewhat rhetorical question, but how would you propose maintaining the
listing of development dependencies?

Also, Sphinx takes literally just seconds to install on most systems. Do
you have a conflict? What's your use case? Just curious here, not attacking
the request.

On Wed, Jul 8, 2015 at 12:05 PM, Christopher Antila <
notifications@github.com> wrote:

> When someone installs Abjad with pip, they install all of the
> dependencies listed in the install_requires variable in setup.py. This
> includes Sphinx, which is used to build documentation from program code,
> and pytest, which is used to run the automated test suite. While pytest
> is quite small, Sphinx is a large package and it requires several
> large-ish dependencies.
>
> Since these packages are (or should be) needed only for Abjad development,
> I suggest removing them from the install_requires argument, relieving
> others from having to install this software that will go unused.
>
> Of course, I may be missing something, in which case I'd be happy to
> learn. Thanks!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Abjad/abjad/issues/468>.
>



-- 
Josiah Wolf Oberholtzer, PhD
http://josiahwolfoberholtzer.com

**Comment by crantila** [comment]
I think it's common to use ``requirements.txt`` files with ``pip`` for managing the development dependencies. You're also already using ``tox``, which does well with that. If you prefer a solution in ``setup.py`` (which is a reasonable place to put the solution) it's also [possible to specify an "extras_require" parameter](https://pythonhosted.org/setuptools/setuptools.html#new-and-changed-setup-keywords) to ``setuptools``. So you might add this...

    setuptools.setup(name='Abjad',
                     ...other stuff...
                     extras_require={'development': ['sphinx', 'pytest']})

In terms of why/when this may be a problem, well, I have to admit it's never a big deal, and only sometimes a small deal. There are two use cases I can imagine. One being for someone running Abjad on a Web server&mdash;The Responsible System Administrator(TM) must choose between removing Abjad's development-only dependencies on every upgrade, or managing updates to software that's effectively unused. The other case hits a little closer to home, imagining the situation where users must download Abjad and its dependencies for a Web app, in which case every extra bit to download will delay users' ability to use the app.

Plus, to put on another hat for a moment, downloading extra software&mdash;no matter how large&mdash;is only "not a problem" for people for whom downloading in general is "not a problem." Sure, Abjad itself is already 25MB in v2.16, so 2MB in dependencies might not make much of a difference, but it is 2MB more than strictly necessary.

Again, this isn't a big problem at all. But in this case, the difficulty of the solution seems proportional to the problematicness of the problem.

**Comment by josiah-wolf-oberholtzer** [comment]
But if you're running Abjad on a web server on the back end (as I do),
you'll almost 100% be running it in a virtual environment where installing
any and all dependencies is not a problem (barring things as complex as
SciPy). That's just standard practice.

As for a downloaded "web app", I'm going to guess you're talking about
something running Python *in* the browser, right? If that's what you're
actually after, there's a whole collection of changes you ought to propose
in tandem. Abjad hits the filesystem a fair amount (for configuration
files, for its pickled LilyPond parser tables, for LilyPond stuff). Our
docs are bundled with the package, and are huge. There are probably other
things that need to be tweaked if you actually expect people to download
Abjad into a browser situation. Let me know what other roadblocks you
encounter here. Might merit some more extensive reworking.

Thanks for the "extra_require" reference. That's compelling.

On Wed, Jul 8, 2015 at 2:26 PM, Christopher Antila <notifications@github.com
> wrote:

> I think it's common to use requirements.txt files with pip for managing
> the development dependencies. You're also already using tox, which does
> well with that. If you prefer a solution in setup.py (which is a
> reasonable place to put the solution) it's also possible to specify an
> "extras_require" parameter
> <https://pythonhosted.org/setuptools/setuptools.html#new-and-changed-setup-keywords>
> to setuptools. So you might add this...
>
> setuptools.setup(name='Abjad',
>                  ...other stuff...
>                  extras_require={'development': ['sphinx', 'pytest']})
>
> In terms of why/when this may be a problem, well, I have to admit it's
> never a big deal, and only sometimes a small deal. There are two use cases
> I can imagine. One being for someone running Abjad on a Web server—The
> Responsible System Administrator(TM) must choose between removing Abjad's
> development-only dependencies on every upgrade, or managing updates to
> software that's effectively unused. The other case hits a little closer to
> home, imagining the situation where users must download Abjad and its
> dependencies for a Web app, in which case every extra bit to download will
> delay users' ability to use the app.
>
> Plus, to put on another hat for a moment, downloading extra software—no
> matter how large—is only "not a problem" for people for whom downloading in
> general is "not a problem." Sure, Abjad itself is already 25MB in v2.16, so
> 2MB in dependencies might not make much of a difference, but it is 2MB more
> than strictly necessary.
>
> Again, this isn't a big problem at all. But in this case, the difficulty
> of the solution seems proportional to the problematicness of the problem.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Abjad/abjad/issues/468#issuecomment-119735136>.
>



-- 
Josiah Wolf Oberholtzer, PhD
http://josiahwolfoberholtzer.com

**Comment by jefftrevino** [comment]
These are exactly the sorts of problems Christopher and I have enumerated in talking about how to get Abjad working in-browser via pypy. I should add to the list that it seems pypy can't do multithreading, which the Abjad code does. 

Is there an accepted way of detecting whether or not a package has been imported into a browser interpreter? It would then be possible to contribute appropriate alternate versions of Abjad code and dynamically run one or the other depending on the situation.

**Comment by crantila** [comment]
In PyPy.js, ``sys.platform`` is set to ``'js'``.

**Comment by josiah-wolf-oberholtzer** [comment]
OK. In the newest version in the repository, pytest and sphinx are no
longer installed by default. To install the development dependencies (in
addition to the others), you need to run "pip install abjad[development]".
That should address Chris's concerns.

I'm not closing this issue yet, because the changes are provisional. I had
to update a fair amount of packages depending on Abjad because of the
change, so I'm still mulling it over.

As for multiprocessing (not multithreading), the *only only* tool in all of
Abjad that can make use of parallelization is the quantizer, and even then
it's purely optional.

On Sat, Jul 11, 2015 at 10:11 PM, Christopher Antila <
notifications@github.com> wrote:

> In PyPy.js, sys.platform is set to 'js'.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Abjad/abjad/issues/468#issuecomment-120687596>.
>



-- 
Josiah Wolf Oberholtzer, PhD
http://josiahwolfoberholtzer.com

**Comment by crantila** [comment]
Thanks. Even if you don't go with this in the end, I do appreciate that you've given this some serious thought and even a demo.

**Comment by josiah-wolf-oberholtzer** [comment]
Well, do make sure that the remaining dependencies work for your purposes.
While we might plausibly remove "configob"j as a dependency, putting
together a replacement based on Python's standard library "configparser"
module, our PLY dependency is definitely not going anywhere. Please double
check that your can use that library in whatever environment you're working
within. Our parsers make extensive use of PLY and also hit the filesystem
for storing and retrieving pickled parser tables.

On Mon, Jul 13, 2015 at 8:15 PM, Christopher Antila <
notifications@github.com> wrote:

> Thanks. Even if you don't go with this in the end, I do appreciate that
> you've given this some serious thought and even a demo.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Abjad/abjad/issues/468#issuecomment-121121162>.
>



-- 
Josiah Wolf Oberholtzer, PhD
http://josiahwolfoberholtzer.com

**Comment by crantila** [comment]
They both appear to be pure Python. Accessing "the filesystem" (held in memory?) isn't an issue, as long as we're shipping the files. In any case, the issues arising there would be quite different from this, so we'll report them as needed.

Regarding @jefftrevino 's earlier comment about the ``multiprocessing`` module, it's actually ``subprocess`` that is causing us problems at the moment (can't import Abjad without it). We'll have to submit a patch for that, but it shouldn't be hard.

**Comment by josiah-wolf-oberholtzer** [comment]
If you're referring to the pickled parser tables, we're not shipping the
files and we never will. So you'll have to propose a patch and tests for
working around that.

I'm still not clear on what environment you're using here, so I can't offer
more detailed advice. If there's any way you can provide some Travis-CI
configuration that duplicates or approaches whatever it is you're trying to
run Abjad under, that would be a big help. That's the only way we're likely
to build in testing for exotic environments.

As for removing subprocess, that's a far bigger request. What exactly are
you proposing instead? Are late imports sufficient? In any case, I'm really
not in the mood to throw late imports or try/excepts into fifty different
locations in our codebase to appease an exotic environment that I can't
even test under. That's a losing battle even if we were to make the change
to get it running. As I mentioned above, if you can provide a Travis-CI
configuration that allows us to test under your environment, I'm willing to
look into this.

On Mon, Jul 13, 2015 at 8:50 PM, Christopher Antila <
notifications@github.com> wrote:

> They both appear to be pure Python. Accessing "the filesystem" (held in
> memory?) isn't an issue, as long as we're shipping the files. In any case,
> the issues arising there would be quite different from this, so we'll
> report them as needed.
>
> Regarding @jefftrevino <https://github.com/jefftrevino> 's earlier
> comment about the multiprocessing module, it's actually subprocess that
> is causing us problems at the moment (can't import Abjad without it). We'll
> have to submit a patch for that, but it shouldn't be hard.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Abjad/abjad/issues/468#issuecomment-121125844>.
>



-- 
Josiah Wolf Oberholtzer, PhD
http://josiahwolfoberholtzer.com

**Comment by josiah-wolf-oberholtzer** [comment]
tl;dr If I can't run an automated full-battery regression test under an
environment equal to or similar to whatever environment you're developing
for, I can't sign off on changes solely targeting that environment.

On Mon, Jul 13, 2015 at 9:06 PM, josiah oberholtzer <
josiah.oberholtzer@gmail.com> wrote:

> If you're referring to the pickled parser tables, we're not shipping the
> files and we never will. So you'll have to propose a patch and tests for
> working around that.
>
> I'm still not clear on what environment you're using here, so I can't
> offer more detailed advice. If there's any way you can provide some
> Travis-CI configuration that duplicates or approaches whatever it is you're
> trying to run Abjad under, that would be a big help. That's the only way
> we're likely to build in testing for exotic environments.
>
> As for removing subprocess, that's a far bigger request. What exactly are
> you proposing instead? Are late imports sufficient? In any case, I'm really
> not in the mood to throw late imports or try/excepts into fifty different
> locations in our codebase to appease an exotic environment that I can't
> even test under. That's a losing battle even if we were to make the change
> to get it running. As I mentioned above, if you can provide a Travis-CI
> configuration that allows us to test under your environment, I'm willing to
> look into this.
>
> On Mon, Jul 13, 2015 at 8:50 PM, Christopher Antila <
> notifications@github.com> wrote:
>
>> They both appear to be pure Python. Accessing "the filesystem" (held in
>> memory?) isn't an issue, as long as we're shipping the files. In any case,
>> the issues arising there would be quite different from this, so we'll
>> report them as needed.
>>
>> Regarding @jefftrevino <https://github.com/jefftrevino> 's earlier
>> comment about the multiprocessing module, it's actually subprocess that
>> is causing us problems at the moment (can't import Abjad without it). We'll
>> have to submit a patch for that, but it shouldn't be hard.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/Abjad/abjad/issues/468#issuecomment-121125844>.
>>
>
>
>
> --
> Josiah Wolf Oberholtzer, PhD
> http://josiahwolfoberholtzer.com
>



-- 
Josiah Wolf Oberholtzer, PhD
http://josiahwolfoberholtzer.com

**Comment by crantila** [comment]
> tl;dr If I can't run an automated full-battery regression test under an environment equal to or similar to whatever environment you're developing for, I can't sign off on changes solely targeting that environment.

Nobody's asking, suggesting, or implying you should do anything other than this. The issue report here is about a relatively small change that's independent of any other hare-brained ideas we have about running Abjad.

I do appreciate your proactive stance in thinking about how our [PyPy.js](https://github.com/pypyjs/pypyjs) plans might affect Abjad, but it's too soon to worry about that. We're still at a very early stage, and we have to address simpler concerns, like how to make the ``unittest`` module work, before we can approach something like Abjad with any hope of understanding the issues. We'll be sure to consult you (Abjad developers) as soon as it's reasonable, but at this point we don't really know what problems we'll encounter, so we can't start thinking about automated testing and CI&mdash;aside from agreeing it's necessary.

**Comment by josiah-wolf-oberholtzer** [comment]
Great. Well, good luck with all of that.

On Tue, Jul 14, 2015 at 11:20 AM, Christopher Antila <
notifications@github.com> wrote:

> tl;dr If I can't run an automated full-battery regression test under an
> environment equal to or similar to whatever environment you're developing
> for, I can't sign off on changes solely targeting that environment.
>
> Nobody's asking, suggesting, or implying you should do anything other than
> this. The issue report here is about a relatively small change that's
> independent of any other hare-brained ideas we have about running Abjad.
>
> I do appreciate your proactive stance in thinking about how our PyPy.js
> <https://github.com/pypyjs/pypyjs> plans might affect Abjad, but it's too
> soon to worry about that. We're still at a very early stage, and we have to
> address simpler concerns, like how to make the unittest module work,
> before we can approach something like Abjad with any hope of understanding
> the issues. We'll be sure to consult you (Abjad developers) as soon as it's
> reasonable, but at this point we don't really know what problems we'll
> encounter, so we can't start thinking about automated testing and CI—aside
> from agreeing it's necessary.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Abjad/abjad/issues/468#issuecomment-121329289>.
>



-- 
Josiah Wolf Oberholtzer, PhD
http://josiahwolfoberholtzer.com

**Comment by josiah-wolf-oberholtzer** [comment]
Done during 2.17 development.

**Comment by crantila** [comment]
Thanks for this!
